src/Eto.WinForms/CustomControls/TreeDataStore.cs

```diff
@@ -0,0 +1,254 @@
+using System;
+using System.Collections;
+using System.Collections.ObjectModel;
+using System.Collections.Specialized;
+using System.Diagnostics;
+using System.Linq;
+using Eto.Forms;
+
+namespace Eto.CustomControls
+{
+	public class TreeDataStore : ITreeGridStore<ITreeGridItem>, IList, INotifyCollectionChanged
+	{
+		private readonly ObservableCollection<ITreeGridItem> cache;
+		private readonly TreeController controller;
+
+		public TreeDataStore(ITreeHandler handler)
+		{
+			controller = new TreeController(this, handler);
+			cache = new ObservableCollection<ITreeGridItem>();
+			cache.CollectionChanged += (s, e) => OnTriggerCollectionChanged(e);
+		}
+
+		internal void Refresh(bool force = false)
+		{
+			// Reset/Clear
+			if (force)
+			{
+				Debug.WriteLine("Clear");
+				cache.Clear();
+			}
+
+			controller.ResetSections();
+
+			int count = controller.Count;
+			var items = new ITreeGridItem[count];
+			for (int index = 0; index < count; index++)
+			{
+				items[index] = controller[index];
+			}
+
+			for (int index = 0; index < items.Length; index++)
+			{
+				var item = items[index];
+
+				// Add
+				if (index == cache.Count)
+				{
+					Debug.WriteLine($"Add at {index}");
+					cache.Insert(index, item);
+					continue;
+				}
+
+				var cachedItemAtIndex = cache[index];
+				if (ReferenceEquals(item, cachedItemAtIndex))
+				{
+					continue;
+				}
+
+				// Remove
+				if (!items.Contains(cachedItemAtIndex))
+				{
+					Debug.WriteLine($"Remove at {index}");
+					cache.RemoveAt(index);
+				}
+
+				int itemIndexInCache = cache.IndexOf(item);
+
+				// Insert
+				if (itemIndexInCache == -1)
+				{
+					Debug.WriteLine($"Insert at {index}");
+					cache.Insert(index, item);
+					continue;
+				}
+
+				// Move
+				if (itemIndexInCache != index)
+				{
+					Debug.WriteLine($"Move from {itemIndexInCache} to {index}");
+					cache.Move(itemIndexInCache, index);
+				}
+			}
+
+			// Trim
+			while (cache.Count > count)
+			{
+				Debug.WriteLine($"Trim at {cache.Count - 1}");
+				cache.RemoveAt(cache.Count - 1);
+			}
+
+			// Temp: Consistency check
+			for (int index = 0; index < items.Length; index++)
+			{
+				if (!ReferenceEquals(cache[index], items[index]))
+					throw new ApplicationException($"Mismatch at index {index}.");
+			}
+
+			Debug.WriteLine("Refresh completed.");
+		}
+
+		internal void RefreshItem(ITreeGridItem item)
+		{
+			if (item == null) return;
+
+			int row = IndexOf(item);
+			if (row == -1) return;
+
+			cache.RemoveAt(row);
+			cache.Insert(row, item);
+		}
+
+		public event EventHandler<TreeGridViewItemCancelEventArgs> Expanding;
+		public event EventHandler<TreeGridViewItemCancelEventArgs> Collapsing;
+		public event EventHandler<TreeGridViewItemEventArgs> Expanded;
+		public event EventHandler<TreeGridViewItemEventArgs> Collapsed;
+
+		// TODO access modifier
```

Comment:
If we move `CollapseRow()` and `ExpandRow()` here, then these methods can become private.